File: php-src/Sending.php
<?php

namespace kalanis\EmailApi;


/**
 * Class Sending
 * Send mails by correct service
 * Select the only one - first successful one
 */
class Sending implements Interfaces\ISending
{
    const CALL_UNKNOWN = 590;
    const CALL_RUN_DIED = 591;
    const CALL_EXCEPTION = 592;

    protected LocalInfo\ServicesOrdering $servicesIterator;
    protected Interfaces\ILocalInfo $info;

    public function __construct(Interfaces\ILocalInfo $info, LocalInfo\ServicesOrdering $ordering)
    {
        $this->info = $info;
        $this->servicesIterator = $ordering;
    }

    public function canUseService(): bool
    {
        return $this->servicesIterator->canUseService();
    }

    public function systemServiceId(): int
    {
        return static::SERVICE_SYSTEM;
    }

    /**
     * @param Interfaces\IContent $content
     * @param Interfaces\IEmailUser $to
     * @param Interfaces\IEmailUser|null $from
     * @param Interfaces\IEmailUser|null $replyTo
     * @param bool $toDisabled
     * @throws Exceptions\EmailException
     * @return Basics\Result
     */
    public function sendEmail(Interfaces\IContent $content, Interfaces\IEmailUser $to, ?Interfaces\IEmailUser $from = null, ?Interfaces\IEmailUser $replyTo = null, $toDisabled = false): Basics\Result
    {
        $this->info->beforeProcess($content, $to, $from);

        if ($this->canUseService()) {
            foreach ($this->servicesIterator as $index => $lib) {
                /** @var Interfaces\ISending $lib */
                if (!$this->isAllowed($lib)) {
                    continue;
                }
                $this->info->beforeSend($lib, $content);
                try {
                    $result = $lib->sendEmail($content, $to, $from, $replyTo, $toDisabled);
                    if ($result->getStatus()) {
                        $this->info->whenResultIsSuccessful($lib, $result);
                        return $result;
                    } else {
                        $this->info->whenResultIsNotSuccessful($lib, $result);
                        $this->servicesIterator->removeService($lib); // throw it out, if we send more mail, then it won't be bothered anymore on this run
                        if ($this->servicesIterator->isReturningAfterFirstUnsuccessful()) {
                            return $result;
                        }
                    }
                } catch (Exceptions\EmailException $ex) {
                    $this->info->whenSendFails($lib, $ex);
                }
            }
        }
        $this->info->whenNoDefinitionIsUsable();
        return new Basics\Result(false, $this->info->getLangSendingFailed());
    }

    /**
     * @param mixed $lib
     * @return bool
     */
    protected function isAllowed($lib): bool
    {
        return (($lib instanceof Interfaces\ISending) && $lib->canUseService());
    }
}

File: kw_email/sending.py
from .exceptions import EmailException
from .interfaces import IContent, IEmailUser, ISending, ILocalInfo
from .local_info import ServicesOrdering
from .result import Result


class Sending(ISending):
    """
     * Send mails by correct service
     * Select the only one - first successful one
    """

    CALL_UNKNOWN = 590
    CALL_RUN_DIED = 591
    CALL_EXCEPTION = 592

    def __init__(self, info: ILocalInfo, ordering: ServicesOrdering):
        self._info = info
        self._services_iterator = ordering

    def can_use_service(self) -> bool:
        return self._services_iterator.can_use_service()

    def system_service_id(self) -> int:
        return self.SERVICE_SYSTEM

    def send_email(self, content: IContent, target: IEmailUser, source: IEmailUser = None, reply_to: IEmailUser = None, to_disabled: bool = False) -> Result:
        self._info.before_process(content, target, source)

        if self.can_use_service():
            for lib in self._services_iterator:
                if not Sending._is_allowed(lib):
                    continue

                result = None
                self._info.before_send(lib, content)
                try:
                    result = lib.send_email(content, target, source, reply_to, to_disabled)
                    if result.get_status():
                        self._info.when_result_is_successful(lib, result)
                        return result
                    else :
                        self._info.when_result_is_not_successful(lib, result)
                        self._services_iterator.remove_service(lib)  # throw it out, if we send more mail, then it won't be bothered anymore on this run
                        if self._services_iterator.is_returning_after_first_unsuccessful():
                            return result

                except EmailException as ex:
                    self._info.when_send_fails(lib, ex)

        self._info.when_no_definition_is_usable()
        return Result(False, self._info.get_lang_sending_failed())

    @staticmethod
    def _is_allowed(lib) -> bool:
        return isinstance(lib, ISending) and lib.can_use_service()
